handler/src/main/java/io/netty/handler/ssl/TlsMasterKeyHandler.java

```diff
@@ -0,0 +1,186 @@
+        return UNAVAILABILITY_CAUSE;
+    }
+
+    /* Returns {@code true} if and only if sun.security.ssl.SSLSessionImpl exists in the runtime.
+     */
+    public static boolean isSunSslEngineAvailable() {
+        return UNAVAILABILITY_CAUSE == null;
+    }
+
+    /**
+     * Consume the master key for the session and the sessionId
+     * @param masterKey A 48-byte secret shared between the client and server.
+     * @param sessionId The sessionId An arbitrary byte sequence chosen by the server to identify an
+     *                  active or resumable session state.
+     */
+    protected abstract void accept(SecretKey masterKey, byte[] sessionId);
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+        //only try to log the session info if the ssl handshake has successfully completed.
+        if (evt == SslHandshakeCompletionEvent.SUCCESS) {
+
+            boolean shouldHandle = SystemPropertyUtil.getBoolean(SYSTEM_PROP_KEY, false);
+
+            //if the handler is not an instance of DoNothingTlsMasterKeyBiConsumer try to handle it
+            if (shouldHandle) {
+                final SslHandler handler = ctx.pipeline().get(SslHandler.class);
+                final SSLEngine engine = handler.engine();
+                final SSLSession sslSession = engine.getSession();
+
+                try {
+                    //the OracleJVM does not expose a way to get the master secret, so try to use reflection to get it.
+                    if (isSunSslEngineAvailable() && sslSession.getClass().equals(SSL_SESSIONIMPL_CLASS)) {
+                        SecretKey secretKey = (SecretKey) SSL_SESSIONIMPL_MASTER_SECRET_FIELD.get(sslSession);
+                        accept(secretKey, sslSession.getId());
+                    } else if (OpenSsl.isAvailable() && engine instanceof OpenSslEngine) {
+                        SecretKeySpec secretKey = new SecretKeySpec(
+                                SSL.getMasterKey(((OpenSslEngine) engine).sslPointer()), "AES");
+                        accept(secretKey, sslSession.getId());
+                    }
+                } catch (IllegalAccessException e) {
+                    throw new IllegalArgumentException("Failed to access the field 'masterSecret' via reflection.", e);
+                }
+            }
+        }
+
+        ctx.fireUserEventTriggered(evt);
+    }
+
+    /**
+     * Record the session identifier and master key to the {@link InternalLogger} named <code>io.netty.wireshark</code>.
+     * ex. <code>RSA Session-ID:XXX Master-Key:YYY</code>
+     * This format is understood by Wireshark 1.6.0.
+     * https://code.wireshark.org/review/gitweb?p=wireshark.git;a=commit;h=686d4cabb41185591c361f9ec6b709034317144b
+     * The key and session identifier are forwarded to the log named 'io.netty.wireshark'.
+     */
+    public static final class WiresharkTlsMasterKeyHandler extends TlsMasterKeyHandler {
```

Comment:
I think I would mark this private and just expose a static method in the abstract class to obtain this handler.